### PR TITLE
Add diffusion kernel to PerfSuite

### PR DESCRIPTION
**Pull request by artv3**
HO FEM which stresses shared memory usage. 

Notes: my flop count slightly over estimates what nvprof reports. But not sure where the difference is....

### CODE REVIEW

File: src/apps/DIFFUSION3DPA-Cuda.cpp [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+#define DIFFUSION3DPA_DATA_TEARDOWN_CUDA                            \
+  getCudaDeviceData(m_Y, Y, DPA_D1D *DPA_D1D *DPA_D1D *m_NE);       \
+  deallocCudaDeviceData(Basis);                                     \
+  deallocCudaDeviceData(dBasis);                                    \
+  deallocCudaDeviceData(D);                                         \
+  deallocCudaDeviceData(X);                                         \
+  deallocCudaDeviceData(Y);
+
+//#define USE_RAJA_UNROLL
```
Comment (rhornung67) [inline src/apps/DIFFUSION3DPA-Cuda.cpp:37]: These macro definitions are repeated in other kernel files (MASS3DPA) and in this PR. Let's move them into one header file and include that where we need them.

File: src/apps/DIFFUSION3DPA-OMP.cpp [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+#include "DIFFUSION3DPA.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#include <iostream>
+
+namespace rajaperf {
+namespace apps {
+
+//#define USE_RAJA_UNROLL
```
Comment (rhornung67) [inline src/apps/DIFFUSION3DPA-OMP.cpp:18]: similar comment on the macros.

**Review by rhornung67 — CHANGES_REQUESTED** [review]
Everything looks good except for the repeated macros.

**Comment by artv3** [comment]
> Everything looks good except for the repeated macros.

Thanks! I cleaned up the macros and moved them to FEM_MACROS.hpp.

**Review by rhornung67 — APPROVED** [review]
Thanks for the macros reorg. Looks good. 

@MrBurmark wanna take a pass through this before merging?

**Comment by artv3** [comment]
Just pushed a small rework of the macros that should make hipcc/clang happy.

**Review by rhornung67 — APPROVED** [review]

File: src/apps/DIFFUSION3DPA-Hip.cpp [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+#define DIFFUSION3DPA_DATA_TEARDOWN_HIP                                   \
+  getHipDeviceData(m_Y, Y, DPA_D1D *DPA_D1D *DPA_D1D *m_NE);              \
+  deallocHipDeviceData(Basis);                                            \
+  deallocHipDeviceData(dBasis);                                           \
+  deallocHipDeviceData(D);                                                \
+  deallocHipDeviceData(X);                                                \
+  deallocHipDeviceData(Y);
+
+// Uncomment to add compiler directives for loop unrolling
+//#define USE_RAJA_UNROLL
```
Comment (MrBurmark) [inline src/apps/DIFFUSION3DPA-Hip.cpp:38]: Why is this still here, wouldn't it need to be above the include to work?

Comment (artv3) [reply src/apps/DIFFUSION3DPA-Hip.cpp:38]: hey good catch! sure let me move up.

**Review by MrBurmark — APPROVED** [review]

**Pull request merged**